pandas/_libs/reduction.pyx

```diff
@@ -548,14 +550,19 @@ def apply_frame_axis0(object frame, object f, object names,
             # Need to infer if low level index slider will cause segfaults
             require_slow_apply = i == 0 and piece is chunk
             try:
-                if piece.index is chunk.index:
-                    piece = piece.copy(deep='all')
-                else:
+                if piece.index is not chunk.index:
                     mutated = True
             except AttributeError:
                 # `piece` might not have an index, could be e.g. an int
                 pass
 
+            if not is_scalar(piece):
+                # Need to copy data to avoid appending references
+                if hasattr(piece, "copy"):
+                    piece = piece.copy(deep="all")
+                else:
+                    piece = copy(piece)
```

Comment:
I think you can *just* call ``copy(piece)`` here, no?

Reply:
I had tried that originally but was getting some odd test failures elsewhere in the code base, not quite sure what the issue was

Reply:
ok let’s try to track that down then

Reply:
@dsaxton any update on that? I think we are close on this PR

Reply:
> @dsaxton any update on that? I think we are close on this PR

One of the issues is with this test: https://github.com/pandas-dev/pandas/blob/master/pandas/tests/groupby/test_groupby.py#L944

If we just try a straight `copy` then we get this output:

```python
[ins] In [4]:     df = DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)}) 
         ...:  
         ...:     result1 = df.groupby("key", group_keys=True).apply(lambda x: x[:].key) 
         ...:     result2 = df.groupby("key", group_keys=True).apply(lambda x: x.key)                                                                                             

[ins] In [5]: result1                                                                                                                                                             
Out[5]: 
key   
1    0    1
     1    1
     2    1
2    3    2
     4    2
     5    2
3    6    3
     7    3
     8    3
Name: key, dtype: int64

[ins] In [6]: result2                                                                                                                                                             
Out[6]: 
key  0  1  2
key         
1    1  1  1
2    2  2  2
3    3  3  3
```

To be honest I'm not really sure what's going on here.

Reply:
Pretty strange. Is that the only failure?

Reply:
There are actually four test failures (at least in `/tests/groupby/`), full output is below:

<details>
<summary>Test output</summary>

```
==================================================================================== FAILURES ====================================================================================
_________________________________________________________________________ test_no_mutate_but_looks_like __________________________________________________________________________

    def test_no_mutate_but_looks_like():
    
        # GH 8467
        # first show's mutation indicator
        # second does not, but should yield the same results
        df = DataFrame({"key": [1, 1, 1, 2, 2, 2, 3, 3, 3], "value": range(9)})
    
        result1 = df.groupby("key", group_keys=True).apply(lambda x: x[:].key)
        result2 = df.groupby("key", group_keys=True).apply(lambda x: x.key)
>       assert_series_equal(result1, result2)

test_groupby.py:956: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

left = key   
1    0    1
     1    1
     2    1
2    3    2
     4    2
     5    2
3    6    3
     7    3
     8    3
Name: key, dtype: int64
right = key  0  1  2
key         
1    1  1  1
2    2  2  2
3    3  3  3, cls = <class 'pandas.core.series.Series'>

    def _check_isinstance(left, right, cls):
        """
        Helper method for our assert_* methods that ensures that
        the two objects being compared have the right type before
        proceeding with the comparison.
    
        Parameters
        ----------
        left : The first object being compared.
        right : The second object being compared.
        cls : The class type to check against.
    
        Raises
        ------
        AssertionError : Either `left` or `right` is not an instance of `cls`.
        """
    
        err_msg = "{name} Expected type {exp_type}, found {act_type} instead"
        cls_name = cls.__name__
    
        if not isinstance(left, cls):
            raise AssertionError(
                err_msg.format(name=cls_name, exp_type=cls, act_type=type(left))
            )
        if not isinstance(right, cls):
            raise AssertionError(
>               err_msg.format(name=cls_name, exp_type=cls, act_type=type(right))
            )
E           AssertionError: Series Expected type <class 'pandas.core.series.Series'>, found <class 'pandas.core.frame.DataFrame'> instead

../../util/testing.py:393: AssertionError
______________________________________________________________________________ test_transform_axis _______________________________________________________________________________

tsframe =                    A         B         C         D
2000-01-03 -0.067269  0.149908 -1.327568 -1.372986
2000-01-04 -1.34....122833  0.392259
2000-02-10 -0.765270 -0.621285 -0.145734  0.185360
2000-02-11 -0.515518  0.155501 -1.637690 -0.447299

    def test_transform_axis(tsframe):
    
        # make sure that we are setting the axes
        # correctly when on axis=0 or 1
        # in the presence of a non-monotonic indexer
        # GH12713
    
        base = tsframe.iloc[0:5]
        r = len(base.index)
        c = len(base.columns)
        tso = DataFrame(
            np.random.randn(r, c), index=base.index, columns=base.columns, dtype="float64"
        )
        # monotonic
        ts = tso
        grouped = ts.groupby(lambda x: x.weekday())
        result = ts - grouped.transform("mean")
>       expected = grouped.apply(lambda x: x - x.mean())

test_transform.py:181: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../core/groupby/groupby.py:726: in apply
    result = self._python_apply_general(f)
../../core/groupby/groupby.py:745: in _python_apply_general
    keys, values, not_indexed_same=mutated or self.mutated
../../core/groupby/generic.py:1159: in _wrap_applied_output
    return self._concat_objects(keys, values, not_indexed_same=not_indexed_same)
../../core/groupby/groupby.py:954: in _concat_objects
    result = result.reindex(ax, axis=self.axis)
../../util/_decorators.py:235: in wrapper
    return func(*args, **kwargs)
../../core/frame.py:3889: in reindex
    return super().reindex(**kwargs)
../../core/generic.py:4599: in reindex
    axes, level, limit, tolerance, method, fill_value, copy
../../core/frame.py:3777: in _reindex_axes
    index, method, copy, level, fill_value, limit, tolerance
../../core/frame.py:3799: in _reindex_index
    allow_dups=False,
../../core/generic.py:4662: in _reindex_with_indexers
    copy=copy,
../../core/internals/managers.py:1247: in reindex_indexer
    self.axes[axis]._can_reindex(indexer)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = DatetimeIndex(['2000-01-07', '2000-01-07', '2000-01-07', '2000-01-07',
               '2000-01-07'],
              dtype='datetime64[ns]', freq='B')
indexer = array([-1, -1, -1, -1,  0,  1,  2,  3,  4])

    def _can_reindex(self, indexer):
        """
        Check if we are allowing reindexing with this particular indexer.
    
        Parameters
        ----------
        indexer : an integer indexer
    
        Raises
        ------
        ValueError if its a duplicate axis
        """
    
        # trying to reindex on an axis with duplicates
        if not self.is_unique and len(indexer):
>           raise ValueError("cannot reindex from a duplicate axis")
E           ValueError: cannot reindex from a duplicate axis

../../core/indexes/base.py:3279: ValueError
____________________________________________________________________________ test_dispatch_transform _____________________________________________________________________________

tsframe =                    A         B         C         D
2000-01-03 -1.100038  1.084505 -0.110031  1.207958
2000-01-04 -0.62....396718 -0.787098
2000-02-10  0.287868 -0.340750  0.240605  2.051640
2000-02-11 -0.384174  1.275258 -0.477559  0.907368

    def test_dispatch_transform(tsframe):
        df = tsframe[::5].reindex(tsframe.index)
    
        grouped = df.groupby(lambda x: x.month)
    
>       filled = grouped.fillna(method="pad")

test_transform.py:315: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../core/groupby/groupby.py:628: in wrapper
    return self.apply(curried)
../../core/groupby/groupby.py:726: in apply
    result = self._python_apply_general(f)
../../core/groupby/groupby.py:745: in _python_apply_general
    keys, values, not_indexed_same=mutated or self.mutated
../../core/groupby/generic.py:1159: in _wrap_applied_output
    return self._concat_objects(keys, values, not_indexed_same=not_indexed_same)
../../core/groupby/groupby.py:936: in _concat_objects
    result = concat(values, axis=self.axis)
../../core/reshape/concat.py:256: in concat
    return op.get_result()
../../core/reshape/concat.py:471: in get_result
    mgrs_indexers, self.new_axes, concat_axis=self.axis, copy=self.copy
../../core/internals/managers.py:2048: in concatenate_block_managers
    return BlockManager(blocks, axes)
../../core/internals/managers.py:143: in __init__
    self._verify_integrity()
../../core/internals/managers.py:345: in _verify_integrity
    construction_error(tot_items, block.shape[1:], self.axes)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tot_items = 4, block_shape = (30,)
axes = [Index(['A', 'B', 'C', 'D'], dtype='object'), DatetimeIndex(['2000-02-01', '2000-02-02', '2000-02-03', '2000-02-04',
 ...000-02-08', '2000-02-09',
               '2000-02-10', '2000-02-11'],
              dtype='datetime64[ns]', freq=None)]
e = None

    def construction_error(tot_items, block_shape, axes, e=None):
        """ raise a helpful message about our construction """
        passed = tuple(map(int, [tot_items] + list(block_shape)))
        # Correcting the user facing error message during dataframe construction
        if len(passed) <= 2:
            passed = passed[::-1]
    
        implied = tuple(len(ax) for ax in axes)
        # Correcting the user facing error message during dataframe construction
        if len(implied) <= 2:
            implied = implied[::-1]
    
        if passed == implied and e is not None:
            raise e
        if block_shape[0] == 0:
            raise ValueError("Empty data passed with indices specified.")
        raise ValueError(
>           "Shape of passed values is {0}, indices imply {1}".format(passed, implied)
        )
E       ValueError: Shape of passed values is (30, 4), indices imply (18, 4)

../../core/internals/managers.py:1708: ValueError
______________________________________________________________________ test_groupby_selection_with_methods _______________________________________________________________________

df =               A      B         C         D
2014-01-01  foo    one -0.258062 -0.381800
2014-01-02  bar    one  0.928351...6  bar    two  0.771773  1.504767
2014-01-07  foo    one -0.786105  2.548049
2014-01-08  foo  three -1.368690  1.008265

    def test_groupby_selection_with_methods(df):
        # some methods which require DatetimeIndex
        rng = date_range("2014", periods=len(df))
        df.index = rng
    
        g = df.groupby(["A"])[["C"]]
        g_exp = df[["C"]].groupby(df["A"])
        # TODO check groupby with > 1 col ?
    
        # methods which are called as .foo()
        methods = [
            "count",
            "corr",
            "cummax",
            "cummin",
            "cumprod",
            "describe",
            "rank",
            "quantile",
            "diff",
            "shift",
            "all",
            "any",
            "idxmin",
            "idxmax",
            "ffill",
            "bfill",
            "pct_change",
            "tshift",
        ]
    
        for m in methods:
>           res = getattr(g, m)()

test_whitelist.py:367: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../core/groupby/groupby.py:628: in wrapper
    return self.apply(curried)
../../core/groupby/groupby.py:726: in apply
    result = self._python_apply_general(f)
../../core/groupby/groupby.py:745: in _python_apply_general
    keys, values, not_indexed_same=mutated or self.mutated
../../core/groupby/generic.py:1159: in _wrap_applied_output
    return self._concat_objects(keys, values, not_indexed_same=not_indexed_same)
../../core/groupby/groupby.py:936: in _concat_objects
    result = concat(values, axis=self.axis)
../../core/reshape/concat.py:256: in concat
    return op.get_result()
../../core/reshape/concat.py:471: in get_result
    mgrs_indexers, self.new_axes, concat_axis=self.axis, copy=self.copy
../../core/internals/managers.py:2048: in concatenate_block_managers
    return BlockManager(blocks, axes)
../../core/internals/managers.py:143: in __init__
    self._verify_integrity()
../../core/internals/managers.py:345: in _verify_integrity
    construction_error(tot_items, block.shape[1:], self.axes)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tot_items = 1, block_shape = (8,)
axes = [Index(['C'], dtype='object'), DatetimeIndex(['2014-01-01', '2014-01-03', '2014-01-05', '2014-01-07',
               '...014-01-03', '2014-01-05',
               '2014-01-07', '2014-01-08'],
              dtype='datetime64[ns]', freq=None)]
e = None

    def construction_error(tot_items, block_shape, axes, e=None):
        """ raise a helpful message about our construction """
        passed = tuple(map(int, [tot_items] + list(block_shape)))
        # Correcting the user facing error message during dataframe construction
        if len(passed) <= 2:
            passed = passed[::-1]
    
        implied = tuple(len(ax) for ax in axes)
        # Correcting the user facing error message during dataframe construction
        if len(implied) <= 2:
            implied = implied[::-1]
    
        if passed == implied and e is not None:
            raise e
        if block_shape[0] == 0:
            raise ValueError("Empty data passed with indices specified.")
        raise ValueError(
>           "Shape of passed values is {0}, indices imply {1}".format(passed, implied)
        )
E       ValueError: Shape of passed values is (8, 1), indices imply (10, 1)

../../core/internals/managers.py:1708: ValueError
================================================== 4 failed, 3401 passed, 2 skipped, 22 xfailed, 2 xpassed in 203.24s (0:03:23) ==================================================
```

</details>